### PR TITLE
Update the exchange match logic to align with the latest spec proposal

### LINKED ISSUE #4451: Need a solution to get correct generated SecureSessionHandle for each new connection
#### Problem
I found https://github.com/project-chip/connectedhomeip/pull/3728 broke the end-to-end messaging testing which is temporarily disabled in CI.

With this change, the app does not specify the target NodeID, but relay on the following callback to update the SecureSessionHandle.

void OnNewConnection(chip::SecureSessionHandle session, chip::SecureSessionMgr * mgr) override { 
    mSecureSession = session;
}
We can only specify one delegate of SecureSessionMgr at a time, but there are multiple connections managed by the SecureSessionMgr, how we make sure the SecureSessionMgr can pass the generated SecureSessionHandle to the corresponding target of its connection?

For example, there are multiple objects are set as the delegate of SecureSessionMgr, including DeviceController, ExchangeManager and application itself. The later SetDelegate() will overwrite the previous one.

Echo Application:
gSessionManager.SetDelegate(&gTestSecureSessionMgrDelegate); // Set the echo app as the delegate of SecureSessionMgr.
err = gExchangeManager.Init(&gSessionManager); // Set the ExchangeManager as the delegate of SecureSessionMgr.

The echo application won't received the generated SecureSessionHandle and it will sendMessage to garbage target.

### PR DESCRIPTION
**Pull request by yufengwangca**
#### Problem
What is being fixed?  
* Currently, SecureSessionHandle is used to match exchange for the incoming message in addition to exchange id and node id, but an exchange is not tied to a secure session in the spec. Before app send an message or create an new exchange, there is no way to know the keyid of a SecureSessionHandle up front, right now, we temporarily  hard coded the keyid to 0 to pass the exchange match check.

After discussion, we agree to update spec with the following changes.
1. reserve sessionId 0 for unsecured sessions
2. clearly specify the Session ID is one of the metadata associated with an Exchange.

* Fixes #4451

#### Change overview
1. Use sessionId 0 for unsecured session
2. Match incoming message's source Session ID with the Session ID associated with the exchange during the exchange match check.

#### Testing
How was this tested? (at least one bullet point required)
1. ./chip-echo-responder
2.  ./chip-echo-requester <ip>
confirm the echo test pass with this change.
1. ./chip-im-responder
2.  ./chip-im-initiator <ip>
confirm the im test pass with this change.

### CODE REVIEW

File: src/messaging/ExchangeContext.cpp
Comment (kghost) [inline src/messaging/ExchangeContext.cpp:300]: This leaves a serious secure hold that a malicious attacker can inject a unsecure packet to a secure session.

**Review by kghost — CHANGES_REQUESTED** [review]

File: src/messaging/ExchangeContext.cpp
Comment (yufengwangca) [reply src/messaging/ExchangeContext.cpp:300]: As we discussed offline, the exchange is not associated with session per spec, and the keyID within session here is not correct (hard coded to 0 to pass the exchange match).

We might need to come up with another solution out of spec to address this issue. Based on spec, when matching an exchange, we should only match source {fabricId, nodeId} with the exchange's desired {fabricId, nodeId} beside exchangeID.

Comment (yufengwangca) [reply src/messaging/ExchangeContext.cpp:300]: The message arrived in exchange layer has already been decoded, if the hacker inject a secure packet with wrong keyID or insecure message to a secure session, it will fail in decryption logic if the KeyID is wrong or the packet is not encrypted using the correct/corresponding key.

**Comment by kghost** [comment]
Discussed with Boris and Martin, actions should be taken on the spec side first.

**Comment by yufengwangca** [comment]
@turon could you confirm if the change in this PR aligned with your proposal in spec?

File: src/messaging/ExchangeContext.cpp
Comment (msandstedt) [reply src/messaging/ExchangeContext.cpp:300]: > the exchange is not associated with session per spec

How can the above be true?  A message incoming over CASE must be matched to the session ID (local key ID) before forwarding to other layers, or the security model evaporates.  The code is not free to forward messages anywhere it pleases just  because the message got in the door.  Exchanges and application layer behavior do have specific associations with sessions and can only receive messages from their associated session.  This means the session ID needs to be associated with the exchange and checked, and if it doesn't match, the message should be discarded.

Without out this, you can maliciously inject messages into code paths that are intended for other peers.  This is a vector for permissions escalation.

Comment (yufengwangca) [reply src/messaging/ExchangeContext.cpp:300]: "An Exchange SHALL be bound to exactly one underlying session" was removed from the spec two month ago. What we need to decide is if an exchange is bonded to its underling session. Such as in CASE resumption case, do we need to drop the old exchange and create a new one, or just update the session.

https://github.com/CHIP-Specifications/connectedhomeip-spec/pull/3588 makes it clear. The exchange is bonded to the underlying session by Session ID.

Comment (msandstedt) [reply src/messaging/ExchangeContext.cpp:300]: If we decide to allow exchanges to persist beyond the lifetime of a session, we would still need to verify that the exchange is associated with the _current_ session.  This is the only way to carry our peer authentication model through the exchange layer.

Comment (msandstedt) [reply src/messaging/ExchangeContext.cpp:300]: And come to think of it, I don't think that's really sufficient.  It does need to be the same session.  In my view, there is an abstract owner / peer in the exchange and this is identified by the session.  It's not just sufficient to authenticate the node.  What if the node is supporting multiple concurrent sessions with a peer?  This becomes nonsensical if an exchange can move to a different session ID.

Comment (yufengwangca) [reply src/messaging/ExchangeContext.cpp:300]: Clarified with Martin, there is only one secure session at a time per peer node within the the same fabric, all exchanges shared the same underling session, and when the session ends, we also need to tear down the associated exchange.
The idea is to make the policy simple in the spec to avoid various confusion in implementation.

Comment (msandstedt) [reply src/messaging/ExchangeContext.cpp:300]: This all makes sense except for:

```
there is only one secure session at a time per peer node within the the same fabric
```

Why?  That's a very significant constraint.

Comment (yufengwangca) [reply src/messaging/ExchangeContext.cpp:300]: Per Martin, the goal is to simply processing of exchanges and bind them to exactly one session and transport. The only use case for muti-session is mutil-fabric, different credentials even to the same peer. For any fabric, the hierarchy of exchanges associating with one underlying session holds, They would associate with the sessions that matches the op cred being used for the interaction as per the ACL.

Do we have any scenario we need to use more than one secure sessions to the same peer node under the same fabric?

Comment (msandstedt) [reply src/messaging/ExchangeContext.cpp:300]: What assumption is leading to the belief that associating an exchange to a single session means there can only be a single session at any given time between two given peers on a given fabric?  This constraint isn't necessary and it is very restrictive.

> Per Martin, the goal is to simply processing of exchanges and bind them to exactly one session and transport.

Great, that's simple and makes perfect sense.

> The only use case for muti-session is mutil-fabric, different credentials even to the same peer.

I disagree.  Why should the case of a single node opening two sessions to a given peer be any different than two separate nodes on the same fabric opening sessions to the given peer?  Nothing in the spec makes such a distinction necessary.  Each node manages its session ID space so it can properly associate and authenticate incoming messages with the correct internal context.

> For any fabric, the hierarchy of exchanges associating with one underlying session holds

Yes, I agree.  Now why can't I then have 10 concurrent exchanges mapping back to 10 sessions between the two peers?  The only thing that should prevent this is lack of capability / memory in one, or the other, or both peers.

Is it because there is an assumption that exchanges also somehow have node IDs as primary key?  You see, that there is the problem.  The whole point of the session ID is that there is an abstraction for nodes to recover the correct context from incoming messages, whatever that might be, and be able to authenticate the association between the incoming message and an internal context.

> They would associate with the sessions that matches the op cred being used for the interaction as per the ACL.

Yes, the session ID maps back to an encryption key, which in turn maps back to OpCreds.  How does this preclude two given peers from opening multiple sessions with one another?

> Do we have any scenario we need to use more than one secure sessions to the same peer node under the same fabric?

Yes, many and they are unknowable.  Currently, the spec can support this and the sdk ought to too.

**Comment by kghost** [comment]
The new patch resolves my concern, but I still think comparing `mSecureSession` is the right way.

I'm considering remove all `packetHeader` and `peerAddress` args above transport layer, because it is not trust-able, and shouldn't be used above transport layer.

File: src/messaging/ExchangeContext.cpp
Comment (yufengwangca) [reply src/messaging/ExchangeContext.cpp:300]: The change in this PR is based on assumption that an exchange is only bind to exactly one underlying session on a given fabric.

Regarding your question, "if we could open more than one session to peer node on a given fabric", I think there is more about policy definition than technical restriction and this is not disabled by this PR. Currently, this is not supported by our SDK. I hope we could clarify this explicitly from the spec side to avoid confusion in implementation.

Comment (msandstedt) [reply src/messaging/ExchangeContext.cpp:300]: I just do not believe there is any ambiguity in this regard within the spec.  The spec is very clear that the session ID is what apps should use to recover context for incoming traffic. This implicitly allows for the multiple concurrent sessions between peers that I'm arguing should be supported.

But the security model also depends upon this, and not doing so is a good way to introduce bugs and security issues.  https://github.com/project-chip/connectedhomeip/pull/8441#issuecomment-883364306 alludes to a potential security issue, and #8481 shows us examples of bugs that can occur when we try to use things other than the session ID to locate context.

I am aware that the `chip::Controller::DeviceController` object is not architected to support multiple concurrent sessions with a single peer because it maintains a single session context per peer and because it demuxes exchange traffic by keying on node ID.  This is not necessarily broken from a security perspective because: 1) the controller is currently single-fabric, and 2) the controller uses a single session ID allocator in such a way that there is a 1:1 mapping between session IDs and node IDs.

But I'm trying to be very forceful in my advocacy that we should not preclude this type of usage at these lower levels, for instance within the exchange manager.  If an implementer wants to roll their own controller object that supports multiple concurrent sessions, this ought to be possible.  Code that would preclude such a thing does not make sense and is probably buggy (again, see #8481).

**Comment by msandstedt** [comment]
> The new patch resolves my concern, but I still think comparing `mSecureSession` is the right way.
> 
> I'm considering remove all `packetHeader` and `peerAddress` arguments above transport layer, because it is not trust-able, and shouldn't be used above transport layer.

I agree.  Does this block this PR, or do we want to write some TODO tickets?

When messages come in the door, we have:

* session ID -> by definition authenticated
* source ID -> part of plaintext nonce
* destination ID -> plaintext authenticated

Obviously session ID is the primary key we should be using to recover internal context.  Perhaps source ID and destination ID are also OK to pass to higher levels of the stack.  But allowing higher level code to key on anything else seems like a good way to inadvertently create security vulnerabilities.

**Comment by yufengwangca** [comment]
> > The new patch resolves my concern, but I still think comparing `mSecureSession` is the right way.
> > I'm considering remove all `packetHeader` and `peerAddress` arguments above transport layer, because it is not trust-able, and shouldn't be used above transport layer.
> 
> I agree. Does this block this PR, or do we want to write some TODO tickets?
> 
> When messages come in the door, we have:
> 
> * session ID -> by definition authenticated
> * source ID -> part of plaintext nonce
> * destination ID -> plaintext authenticated
> 
> Obviously session ID is the primary key we should be using to recover internal context. Perhaps source ID and destination ID are also OK to pass to higher levels of the stack. But allowing higher level code to key on anything else seems like a good way to inadvertently create security vulnerabilities.

The spec change(CHIP-Specifications/connectedhomeip-spec#3588) has specified how should we identify an exchange.
`The Exchange is then identified by the tuple {`Session ID`, `Initiator Fabric ID`, `Initiator Node ID`, `Exchange ID`}.`

The logic in the PR is implemented against the spec except comparing the `Initiator Fabric ID` (Fabric logic is not completely ready in transport layer, will add this logic later to fully comply with the spec).

File: src/messaging/ExchangeContext.cpp
Comment (yufengwangca) [reply src/messaging/ExchangeContext.cpp:300]: My point is this PR does not enable or disable multiple session support to a peer on a given fabric, it uses session ID to match the exchange for incoming packet, it seems it is aligned with what you described above, please lets me know if my understanding is wrong.

Comment (msandstedt) [reply src/messaging/ExchangeContext.cpp:300]: Yes, I was specifically arguing against this statement from https://github.com/project-chip/connectedhomeip/pull/8441/files#r671514752:

> there is only one secure session at a time per peer node within the the same fabric

I agree that the code in this PR does not enforce such a thing.

Getting back to the code that is actually in this PR, in fact I think this line is actually incorrect.  The session ID is the local key ID.  Shouldn't this be the following?

```
        // AND The Session ID associated with the incoming message matches the Session ID associated with the exchange.
        && (mSecureSession.GetLocalKeyId() == session.GetLocalKeyId())
```

I guess I also don't know why we have a delta here at all.  I would have thought the `SecureSessionHandle` == operator overload would be what we should be using.  But that also seems to be incorrect:

```
    bool operator==(const SecureSessionHandle & that) const
    {
        return mPeerNodeId == that.mPeerNodeId && mPeerKeyId == that.mPeerKeyId && mAdmin == that.mAdmin;
    }
```

Again, this should be the local key, no?  In other words, we should be matching on the number that came from our local session ID allocator.  That's what we authenticate against.

Comment (yufengwangca) [reply src/messaging/ExchangeContext.cpp:300]: That needs to be further specified in spec, currently, we only have PeerKeyId defined in session handle. Will update the PR as long as it is clarified

Comment (msandstedt) [reply src/messaging/ExchangeContext.cpp:300]: In my reading of the spec, this is clear.  Each message has a session ID field.  This is plain text and tells the recipient which of its keys was used to encrypt the message and, therefore, informs the recipient of which key it should use to decrypt the message.  If the recipient can use this key to decrypt the message and the MIC shows integrity, the recipient has authentication.

I believe in the sdk, this is referred to as 'local key ID'.  The `SecureSessionHandle` does need the peer key to encrypt outgoing traffic, but it appears it's missing the local key ID, which is ultimately the thing that should be used for session matching.

Another way to look at this: I set my session ID (local key ID) and all my peers set their session IDs (peer key IDs).  I can't authenticate based upon peer key ID because I don't set this and it's not guaranteed unique.

**Comment by bzbarsky-apple** [comment]
> When messages come in the door, we have:
>    session ID -> by definition authenticated
>    source ID -> part of plaintext nonce
>    destination ID -> plaintext authenticated

Just as a minor correction: for unicast encrypted messages, spec does not send a source id and destination id at all.  They are recoverable (and recovered) from the session id.

**Comment by msandstedt** [comment]
> > > The new patch resolves my concern, but I still think comparing `mSecureSession` is the right way.
> > > I'm considering remove all `packetHeader` and `peerAddress` arguments above transport layer, because it is not trust-able, and shouldn't be used above transport layer.
> > 
> > 
> > I agree. Does this block this PR, or do we want to write some TODO tickets?
> > When messages come in the door, we have:
> > 
> > * session ID -> by definition authenticated
> > * source ID -> part of plaintext nonce
> > * destination ID -> plaintext authenticated
> > 
> > Obviously session ID is the primary key we should be using to recover internal context. Perhaps source ID and destination ID are also OK to pass to higher levels of the stack. But allowing higher level code to key on anything else seems like a good way to inadvertently create security vulnerabilities.
> 
> The spec change([CHIP-Specifications/connectedhomeip-spec#3588](https://github.com/CHIP-Specifications/connectedhomeip-spec/pull/3588)) has specified how should we identify an exchange.
> `The Exchange is then identified by the tuple {`Session ID`, `Initiator Fabric ID`, `Initiator Node ID`, `Exchange ID`}.`
> 
> The logic in the PR is implemented against the spec except comparing the `Initiator Fabric ID` (Fabric logic is not completely ready in transport layer, will add this logic later to fully comply with the spec).

Yes, this needs to be addressed in ([CHIP-Specifications/connectedhomeip-spec#3588](https://github.com/CHIP-Specifications/connectedhomeip-spec/pull/3588)) as well.  What's proposed there doesn't make sense.

The session ID is an abstraction that, from the perspective of a recipient, can map to a subject.  But the subject may or may not be a fabric+node identity.  In the case of a PASE session, it isn't.  If exchanges can exist over PASE, then keying on fabric+node identity doesn't work.

**Review by msandstedt — CHANGES_REQUESTED** [review]
This code is attempting session match based upon the peer key ID.  That's not guaranteed unique and not secure.

Session ID is the local key ID, and this is what should be used to verify session match.

**Comment by yufengwangca** [comment]
Hold until https://github.com/CHIP-Specifications/connectedhomeip-spec/pull/3588 is resolved

**Comment by woody-apple** [comment]
Removing SDK Discussion required, this doesn't require anything beyond the normal PR approval process.

**Review by msandstedt — APPROVED** [review]

File: src/transport/PeerConnections.h
Comment (pan-apple) [inline src/transport/PeerConnections.h:212]: This function is a bit confusing. Both lookup parameters are `Optional`. What happens if value of both the parameters is not provided? Do we need a more specific set of lookup functions?

Comment (yufengwangca) [reply src/transport/PeerConnections.h:212]: They are both indeed Optional values per spec.
FindPeerConnectionState has already taken care of cases when those values are not provided during lookup nodeId.ValueOr(kUndefinedNodeId), peerKeyId.ValueOr(0);

**Comment by andy31415** [comment]
@kghost  could you take another  look? The 'changes requested' seems to be very old (30 days ago) and we had many changes in since then.

File: src/transport/PeerConnections.h
Comment (msandstedt) [reply src/transport/PeerConnections.h:212]: By the way, `src/transport/PeerConnections.h` is already pretty broken.  See https://github.com/project-chip/connectedhomeip/issues/8481.

Regardless of what we do in this PR, we will need some significant rework there.

**Comment by kghost** [comment]
I'm working with Michael for a solution on this PR.

File: src/transport/PeerConnections.h
Comment (yufengwangca) [reply src/transport/PeerConnections.h:212]: ok, I can keep FindPeerConnectionState unchanged in this PR and leave to #8481 to address the broken issue.

**Comment by msandstedt** [comment]
> I'm working with Michael for a solution on this PR.

Can we merge this though in the meantime?   It is conformant to https://github.com/CHIP-Specifications/connectedhomeip-spec/pull/3706.  I think we can introduce the followup work in other PRs.

This PR fixes a security problem by now properly associating PASE and CASE sessions to the local key ID instead of the peer key ID, so is useful on its own.

**Review by kghost — APPROVED** [review]
It looks fine to me. I'd like to rename `SecureSessionHandle` to `SessionHandle`

**Comment by yufengwangca** [comment]
> It looks fine to me. I'd like to rename `SecureSessionHandle` to `SessionHandle`

Thanks, that is already part of this change.

**Review by pan-apple — APPROVED** [review]

File: src/transport/SessionHandle.h
Comment (andy31415) [inline src/transport/SessionHandle.h:42]: I expect a equality match to use all members in the class. 

Could we use a method here instead? it is not obvious why mFabric is completely ignored and now the local key Id is ignored as well.

File: src/transport/SessionHandle.h
Comment (andy31415) [inline src/transport/SessionHandle.h:44]: foo.match(bar) is not obvious ... match what?

How about 'compatibleLocalKey' or similar to make the code easier to read?

Comment (yufengwangca) [reply src/transport/SessionHandle.h:44]: This method is introduced for match exchange from session, how about matchExchange?

Comment (yufengwangca) [reply src/transport/SessionHandle.h]: Rename to MatchIncomingSession for now, please lets me know if you have a better name

**Review by andy31415 — APPROVED** [review]

**Review by jmartinez-silabs — APPROVED** [review]

**Review by saurabhst — APPROVED** [review]

**Pull request merged**
